step/src/main/java/com/google/sps/MatchManager.java

```diff
@@ -17,50 +17,113 @@
+  private static double MINIMUM_INTERESTS_PERCENTAGE_REQUIRED = 0.6;
+  private static int QUESTIONS = 5;
 
   private MatchManager() {
 
   }
 
   // Getter method for the match queue (for testing purposes)
-  public static Queue<User> getMatchQueue() {
+  public static HashSet<User> getMatchQueue() {
     return matchQueue;
   }
 
   // Getter method to return a deep copy of the match queue (for testing purposes).
   // The returned queue can be modified without the changes effecting the actual match queue.
-  public static Queue<User> getDeepCopyMatchQueue() {
-    return new LinkedList<User>(matchQueue);
+  public static HashSet<User> getDeepCopyMatchQueue() {
+    return new HashSet <User>(matchQueue);
   }
 
-  // FIFO matchmaker -- takes in the user who just requested a match and matches them
-  // with anyone else who is in the queue waiting to be matched or adds them to the
+  // Setter method to add users directly to the matchqueue (for testing purposes).
+  public static void addUserToMatchQueue(User u) {
+      matchQueue.add(u);
+  }
+
+  // Matchmaker -- takes in the user who just requested a match and matches them
+  // with the one who is in the queue having the highest number of mutual interests with the person
+  // or adds them to the
   // queue to wait to be matched if no on else is in there
   public static void generateMatch(User user) {
     user.updateMatchPendingStatus(true);
 
     if (matchQueue.isEmpty()) {
-      matchQueue.add(user);  
-    } else {
-      User matchResult = matchQueue.peek();
+        matchQueue.add(user); 
+        return;
+    }
+    matchQueue.add(user); 
+
+    if (!matchQueue.isEmpty()) {
+      matchQueue.remove(user);
+      User matchResult = findCompatibleMatch(user);
+      if (matchResult == null) {
+          matchQueue.add(user); 
+          user.updateMatchPendingStatus(true);
+          return;
+      }
 
       if (!matchResult.isMatchedWith(user) && !matchResult.equals(user)) {
         // actually removing the match result from the queue
-        matchResult = matchQueue.poll(); 
         createMatch(matchResult, user);
+        matchQueue.remove(matchResult);  
+        
       // adding the user to the queue in case there wasn't a successful match for them
       // this is so they can be matched later      
-      } else if (!matchQueue.contains(user)) {
-        matchQueue.add(user);  
+      } else if (!matchQueue.contains(user)) {  
+          matchQueue.add(user); 
       }  
     }    
   }
 
+  // Match the user with the most compatible user in the database (the one who has the most number
+  // of similar interests). 
+  // Currently, the minimum percentage of mutual interests between 2 users required to get matched are 60%
+  public static User findCompatibleMatch(User firstUser) {
+    HashMap<User, Integer> mutualInterests = new HashMap<>();
+    for (User secondUser: matchQueue) {
+        mutualInterests.put(secondUser, 0);
```

Comment:
You could probably move this line to be the first line after what is currently line 96